src/wp-includes/media.php

```diff
@@ -1073,14 +1072,14 @@ function wp_get_attachment_image( $attachment_id, $size = 'thumbnail', $icon = f
 			$context
 		);
 
-		// Add loading optimization attributes if not available.
-		$attr = array_merge( $attr, $loading_optimization_attr );
-
-		// Omit the `decoding` attribute if the value is invalid according to the spec.
-		if ( empty( $attr['decoding'] ) || ! in_array( $attr['decoding'], array( 'async', 'sync', 'auto' ), true ) ) {
+		// The `decoding` attribute is now managed within `$loading_optimization_attr`.
+		if ( array_key_exists( 'decoding', $attr ) ) {
 			unset( $attr['decoding'] );
 		}
 
+		// Add loading optimization attributes if not available.
+		$attr = array_merge( $attr, $loading_optimization_attr );
```

Comment:
What is the rationale for this change? As far as I can tell, this should remain as it was before, in order to support force-omitting the `decoding` attribute, e.g. when someone passes `array( 'decoding' => false )` to `wp_get_attachment_image()`.

Reply:
@felixarntz,

The idea was to concentrate on a single place the validation of the `decoding` attribute. The `$attr` sent to `wp_get_attachment_image()` will be used by `wp_get_loading_optimization_attributes()` — the source of truth for the optimization attributes — via the cloned `$loading_attr`.

Unsetting the key from the `$attr` will ensure bad `decoding` values aren't re-added after. 

This is covered by the tests on `test_wp_get_attachment_image_decoding_attr()`, and I don't think this change introduces a regression here.

Reply:
@pereirinha The reason that the `decoding` values from `$attr` needs to be passed to `wp_get_loading_optimization_attributes()` is that that function needs to respect any hard-coded attributes that a developer (or end user via e.g. a plugin) has already added to the image tag, so we shouldn't unconditionally unset said value. That's the primary issue here.

To implement a central validation point, it would probably be best to handle this validation within `wp_get_loading_optimization_attributes()`.

Reply:
@felixarntz 

I believe we are on the same page here. On line `1066`, we duplicate `$attr` into `$loading_attr`. We then send `$loading_attr` to `wp_get_loading_optimization_attributes()` where we respect the value previously set — 3rd party plugin or a dev —, apply the filter `wp_img_tag_add_decoding_attr` and finally do the proper validation.

The reason to unset the `decoding` key of the `$attr` is so that any original value doesn't pollute the processed value.

For instance, say that `wp_get_attachment_image()` is being called by a plugin, which sets a nonsense value for `decoding` like `wrong`. The value will be processed by `wp_get_loading_optimization_attributes()` and removed. The logic we see here on `1076` will ensure that `wrong` isn't set again as the value for the `decoding`. These use cases are covered by these items in the [data provider](https://github.com/WordPress/wordpress-develop/pull/4991/files#diff-ad2d1463dfa406fd36d74f5eb1ba3c0806249b7cdbe6cd969af6bacc99b02af6R3551-R3606) for `wp_get_loading_optimization_attributes()`.

Reply:
@pereirinha Thanks for the detailed explanation. I get what you mean now, and what you're saying makes sense, however there are still two things that don't line up for me here:
* Why do we even unset `$attr['decoding']` the way it's done here? Right afterwards we merge `$attr` with `$loading_optimization_attr`, so then it may be overwritten anyway.
* What I think we _should_ do instead is unset `$attr['decoding']` _after_ merging the attributes if it is `false`, which is an explicit value to omit the attribute, basically enforce omitting the value even if core's default is to provide it. That is similar to what the lines below already do for the `loading` and `fetchpriority` attributes, and we should support the same for `decoding`.